cmd/checkignore.go

```diff
@@ -0,0 +1,234 @@
+	flags.SetOutput(flag.CommandLine.Output())
+	flags.Usage = func() {
+		flag.Usage()
+		fmt.Fprintf(flag.CommandLine.Output(), "\n\nOptions:\n")
+		flags.PrintDefaults()
+	}
+
+	quiet := false
+	flags.BoolVar(&quiet, "quiet", false, "Don't output anything, just set exit status. This is only valid with a single pathname.")
+	flags.BoolVar(&quiet, "q", false, "Alias for --quiet")
+	verbose := false
+	flags.BoolVar(&verbose, "verbose", false, "Also output details about the matching pattern (if any) for each given pathname.")
+	flags.BoolVar(&verbose, "v", false, "Alias for --verbose")
+	nonMatch := false
+	flags.BoolVar(&nonMatch, "non-matching", false, "Show given paths which don’t match any pattern.")
+	flags.BoolVar(&nonMatch, "n", false, "Alias for --non-matching.")
+
+	stdin := false
+	flags.BoolVar(&stdin, "stdin", false, "Read pathnames from the standard input, one per line, instead of from the command-line.")
+
+	noIndex := false
+	flags.BoolVar(&noIndex, "no-index", false, "Don’t look in the index when undertaking the checks.")
+
+	machine := false
+	flags.BoolVar(&machine, "z", false, "The output format is modified to be machine-parseable.")
+
+	flags.Parse(args)
+	args = flags.Args()
+
+	if dir, err := os.Getwd(); err == nil && (dir == c.GitDir.String() || strings.HasPrefix(dir, c.GitDir.String()+"/")) {
+		fmt.Fprintf(flag.CommandLine.Output(), "fatal: This operation must be run in a work tree\n")
+		flags.Usage()
+		os.Exit(128)
+	}
+
+	if machine && !stdin {
+		fmt.Fprintf(flag.CommandLine.Output(), "fatal: -z only makes sense with --stdin\n")
+		flags.Usage()
+		os.Exit(128)
+	}
+
+	if !stdin && len(args) < 1 {
+		fmt.Fprintf(flag.CommandLine.Output(), "fatal: no path specified\n")
+		flags.Usage()
+		os.Exit(128)
+	} else if stdin && len(args) > 0 {
+		fmt.Fprintf(flag.CommandLine.Output(), "fatal: cannot specify pathnames with --stdin\n")
+		flags.Usage()
+		os.Exit(128)
+	} else if quiet && len(args) != 1 && !stdin {
+		fmt.Fprintf(flag.CommandLine.Output(), "fatal: --quiet is only valid with a single pathname\n")
+		flags.Usage()
+		os.Exit(128)
+	} else if quiet && verbose {
+		fmt.Fprintf(flag.CommandLine.Output(), "fatal: cannot have both --quiet and --verbose\n")
+		flags.Usage()
+		os.Exit(128)
+	}
+
+	if !stdin {
+		paths := make([]git.File, 0, len(args))
+
+		// Assemble the list of files, checking for any invalid input
+		for _, p := range args {
+			f := git.File(p)
+			if s, submodule, _ := f.IsInSubmodule(c); s {
+				fmt.Fprintf(os.Stderr, "fatal: Pathspec '%v' is in submodule '%s'\n", p, submodule)
+				os.Exit(128)
+			}
+			if i, _ := f.IsInsideSymlink(); i {
+				fmt.Fprintf(os.Stderr, "fatal: pathspec '%v' is beyond a symbolic link\n", p)
+				os.Exit(128)
+			}
+
+			paths = append(paths, git.File(p))
+		}
+
+		// Invoke the ignore routines directly, rather than relying on
+		//  LsFiles because we need to be able to do things such as
+		//  return non-matches and match details that aren't supported there.
```

Comment:
I don't think I understand this comment.  Isn't determining whether nonmatches or matches are returned what the `-i` option (which admittedly isn't currently implemented, but that's effectively what this PR is doing) toggles? What details that aren't supported need to be matched?

Reply:
Right the "-i" option is not yet enabled and the LsFiles won't return back the pattern that matched or wasn't matched so that checkignore can output that with the -v flag.